src/session.rs

```diff
@@ -0,0 +1,159 @@
+/// act as if it was never found.
+///
+/// The callback should return != 0 to indicate that a value was retrieved
+/// and written in its entirety into `buf`.
+///
+/// When `remove_after` is != 0, the returned data needs to be removed
+/// from the store.
+///
+/// NOTE: the passed in `key` and `buf` are only availabe during the
+/// callback invocation.
+/// NOTE: callbacks used in several sessions via a common config
+/// must be implemented thread-safe.
+pub type rustls_session_store_get_callback = Option<
+    unsafe extern "C" fn(
+        userdata: rustls_session_store_userdata,
+        key: *const rustls_slice_bytes,
+        buf: *mut u8,
+        count: size_t,
+        remove_after: c_int,
+        out_n: *mut size_t,
+    ) -> c_int,
+>;
+
+pub(crate) type SessionStoreGetCallback = unsafe extern "C" fn(
+    userdata: rustls_session_store_userdata,
+    key: *const rustls_slice_bytes,
+    buf: *mut u8,
+    count: size_t,
+    remove_after: c_int,
+    out_n: *mut size_t,
+) -> c_int;
+
+/// Prototype of a callback that can be installed by the application at the
+/// `rustls_server_config` or `rustls_client_config`. This callback will be
+/// invoked by a TLS session when a TLS session has been created and an id
+/// for later use is handed to the client/has been received from the server.
+/// `userdata` will be supplied as provided when registering the callback.
+///
+/// The callback should return != 0 to indicate that the value has been
+/// successfully persisted in its store.
+///
+/// NOTE: the passed in `key` and `val` are only availabe during the
+/// callback invocation.
+/// NOTE: callbacks used in several sessions via a common config
+/// must be implemented thread-safe.
+pub type rustls_session_store_put_callback = Option<
+    unsafe extern "C" fn(
+        userdata: rustls_session_store_userdata,
+        key: *const rustls_slice_bytes,
+        val: *const rustls_slice_bytes,
+    ) -> c_int,
+>;
+
+pub(crate) type SessionStorePutCallback = unsafe extern "C" fn(
+    userdata: rustls_session_store_userdata,
+    key: *const rustls_slice_bytes,
+    val: *const rustls_slice_bytes,
+) -> c_int;
+
+pub(crate) struct SessionStoreBroker {
+    pub userdata: rustls_session_store_userdata,
+    pub get_cb: SessionStoreGetCallback,
+    pub put_cb: SessionStorePutCallback,
+}
+
+impl SessionStoreBroker {
+    pub fn new(
+        userdata: rustls_session_store_userdata,
+        get_cb: SessionStoreGetCallback,
+        put_cb: SessionStorePutCallback,
+    ) -> Self {
+        SessionStoreBroker {
+            userdata,
+            get_cb,
+            put_cb,
+        }
+    }
+
+    fn retrieve(&self, key: &[u8], remove: bool) -> Option<Vec<u8>> {
+        let key: rustls_slice_bytes = key.into();
+        // TODO: we need a buffer where th client can store the retrieved
+        // session value. What size should it have? 10k seems excessive...
+        let mut data: Vec<u8> = vec![0 as u8; 10 * 1024];
```

Comment:
I think `as u8` is not needed here because of the type annotation.

(Some fiddly thoughts below):

For this use case I think we can allocate a fixed-size array on the stack, and then after the callback is done, we can turn that array into a Vec of exactly the right size before returning it.

Now there's the question of: is that better, or worse? Checking the code, it looks like the returned `Vec<u8>` is [immediately treated as a slice to be read into a session](https://github.com/ctz/rustls/blob/e3e626d4dc6f1ffa3ee99e43893582ec5e62ca21/rustls/src/server/tls13.rs#L529-L532), and then dropped.

So, disregard the suggestion about using a fixed-size array. Making a potentially-oversized Vec here is fine because it gets thrown away shortly. And it saves us a copy.